File: server/typescript/src/application/use-cases/delete-item.use-case.ts
import type { IItemRepository } from '../../domain/repositories/item.repository.port.js';
import { NotFoundError } from '../../shared/errors.js';

/**
 * Delete item use case
 */
export class DeleteItemUseCase {
  constructor(private readonly itemRepository: IItemRepository) {}

  /**
   * Execute the use case
   */
  async execute(id: number): Promise<void> {
    // Check if item exists
    const exists = await this.itemRepository.exists(id);
    if (!exists) {
      throw new NotFoundError('Item', id);
    }

    // Delete item from repository
    const deleted = await this.itemRepository.delete(id);

    if (!deleted) {
      throw new NotFoundError('Item', id);
    }
  }
}

File: server/kotlin/src/main/kotlin/com/lebonpoint/application/usecases/DeleteItemUseCase.kt
package com.lebonpoint.application.usecases

import com.lebonpoint.application.services.ValidationService
import com.lebonpoint.domain.repositories.ItemRepository
import com.lebonpoint.shared.NotFoundException

/**
 * Use case for deleting an item
 */
class DeleteItemUseCase(
    private val repository: ItemRepository
) {
    /**
     * Execute the use case
     * @param id Item ID
     * @return Unit if deleted
     * @throws NotFoundException if item not found
     */
    suspend fun execute(id: Int) {
        ValidationService.validateId(id)

        val deleted = repository.delete(id)
        if (!deleted) {
            throw NotFoundException("Item", id)
        }
    }
}
